pkg/filters/pidSet.go

```diff
@@ -20,50 +20,76 @@ import (
 	"github.com/cilium/tetragon/api/v1/tetragon"
 	v1 "github.com/cilium/tetragon/pkg/oldhubble/api/v1"
 	hubbleFilters "github.com/cilium/tetragon/pkg/oldhubble/filters"
+	lru "github.com/hashicorp/golang-lru/v2"
 )
 
-/* pidSet is the set of pids we are interested in receiving events
- * for. This includes the pids listed in filter and any of their
- * children.
- */
-var pidSet map[uint32]bool
+type ChildrenCache = lru.Cache[uint32, struct{}]
 
-func filterByPidSet(pids []uint32) hubbleFilters.FilterFunc {
-	return func(ev *v1.Event) bool {
-		process := GetProcess(ev)
-		if process == nil {
-			return false
-		}
-		for _, pid := range pids {
-			if pid == process.Pid.GetValue() {
-				pidSet[pid] = true
-				return true
-			}
-		}
-		parent := GetParent(ev)
-		if parent == nil {
-			return false
-		}
-		if pidSet[parent.Pid.GetValue()] == true {
+const (
+	childrenCacheSize = 8192
+)
+
+func checkPidSetMembership(pid uint32, pidSet []uint32, childrenCache *ChildrenCache) bool {
+	// Check the original pidSet. The reason for doing this separately is that we never
+	// want to drop the original pidSet from the LRU. Keeping this separately in a slice
+	// is an easy way to achieve this.
+	for _, p := range pidSet {
+		if pid == p {
 			return true
 		}
-		for _, pid := range pids {
-			if pid == parent.Pid.GetValue() {
-				pidSet[pid] = true
-				return true
-			}
-		}
+	}
+	// Fall back to childrenCache to check children.
+	_, ok := childrenCache.Get(pid)
+	return ok
+}
+
+func eventIsInPidSet(ev *v1.Event, pidSet []uint32, childrenCache *ChildrenCache) bool {
+	process := GetProcess(ev)
+	if process == nil {
+		return false
+	}
+
+	// Check the process against our cache
+	pid := process.Pid.GetValue()
+	if checkPidSetMembership(pid, pidSet, childrenCache) {
+		return true
+	}
+
+	parent := GetParent(ev)
+	if parent == nil {
 		return false
 	}
+
+	// Check the parent against our cache
+	ppid := parent.Pid.GetValue()
+	if checkPidSetMembership(ppid, pidSet, childrenCache) {
+		// Add our own PID to the children cache so that we can match our future children
+		childrenCache.Add(pid, struct{}{})
+		return true
+	}
+
+	// No matches, return false
+	return false
+}
+
+func filterByPidSet(pidSet []uint32, childrenCache *ChildrenCache) hubbleFilters.FilterFunc {
+	return func(ev *v1.Event) bool {
+		return eventIsInPidSet(ev, pidSet, childrenCache)
+	}
 }
 
 type PidSetFilter struct{}
 
 func (f *PidSetFilter) OnBuildFilter(_ context.Context, ff *tetragon.Filter) ([]hubbleFilters.FilterFunc, error) {
```

Comment:
Do we need a flag for enabling the children cache? Do we need to track children in all cases?
I was not aware of this filter to be honest, but I'd be surprised if by default it filtered children as well. Especially since the children filtering might have some intricacies (see previous comment.). So let's document the behavior (when we figure it out) and add a flag if we need to.

Reply:
If you don't want to track children then you would simply use the pid filter rather than the pidSet filter. My understanding based on reading the existing implementation was that it was always the intention to have pidSet track children but the implementation was simply not working as intended.

If we do think people might use this in production, then yes I think it definitely makes sense to document the behaviour and potentially provide a flag for configuring cache size. But as I said I think the goal here is to have something useful for debugging/testing and not so much for a production environment.

Reply:
> So my question then is, is there a way for end-users to use this filter?

As of right now, no besides what we expose through GetEventsRequest API. This can change if needed.

Reply:
I think that's OK. Thanks for checking it out!

Reply:
@kkourt I think I misread the question and thought you were asking about whether the cache size would be configurable. The filter could in theory be used through our configmap, same as the others. I can add a check to block this though.